zeebe/gateway-grpc/src/main/java/io/camunda/zeebe/gateway/interceptors/impl/AuthenticationHandler.java

```diff
@@ -75,17 +78,61 @@ public Either<Status, Context> authenticate(final String authorizationHeader) {
       final var username =
           token.getClaims().get(oidcAuthenticationConfiguration.getUsernameClaim());
 
-      if (username == null) {
-        return Either.left(
-            Status.UNAUTHENTICATED.augmentDescription(
-                "Expected a claim '%s' in the token, but it was not present"
-                    .formatted(oidcAuthenticationConfiguration.getUsernameClaim())));
+      if (username != null) {
+        if (username instanceof String) {
+          return Either.right(
+              Context.current()
+                  .withValue(USERNAME, username.toString())
+                  .withValue(USER_CLAIMS, token.getClaims()));
+        } else {
+          return Either.left(
+              Status.UNAUTHENTICATED.augmentDescription(
+                  CONFIGURED_CLAIM_NOT_A_STRING.formatted(
+                      "username", oidcAuthenticationConfiguration.getUsernameClaim())));
+        }
+      }
+
+      final var applicationId =
+          token.getClaims().get(oidcAuthenticationConfiguration.getApplicationIdClaim());
+
+      if (applicationId != null) {
+        if (applicationId instanceof String) {
+          return Either.right(
+              Context.current()
+                  .withValue(APPLICATION_ID, applicationId.toString())
+                  .withValue(USER_CLAIMS, token.getClaims()));
+        } else {
+          return Either.left(
+              Status.UNAUTHENTICATED.augmentDescription(
+                  CONFIGURED_CLAIM_NOT_A_STRING.formatted(
+                      "application id", oidcAuthenticationConfiguration.getApplicationIdClaim())));
+        }
+      }
+
+      return Either.left(
+          Status.UNAUTHENTICATED.augmentDescription(
+              "Expected either a username (claim: %s) or application ID (claim: %s) on the token, but no matching claim found"
+                  .formatted(
+                      oidcAuthenticationConfiguration.getUsernameClaim(),
+                      oidcAuthenticationConfiguration.getApplicationIdClaim())));
+    }
+
+    private Either<Status, String> getApplicationIdClaim(final Map<String, Object> claims) {
```

Comment:
:wrench: This should be unused now, right? If not, it still contains the null + Either anti-pattern.